Dockerfile

```diff
@@ -0,0 +1,18 @@
+FROM golang:1.21-alpine3.17
+
+LABEL mantainer="danieeeld2@correo.ugr.es"
+LABEL version="5.0.0"
+
+WORKDIR /app/test
+
+COPY go.mod go.sum ./
+
+RUN go mod tidy
+
+RUN addgroup -S testgroup && adduser -S -u 1001 test -G testgroup
+
+USER test
+
+COPY internal internal
```

Comment:
¿Dónde estás copiando internal? ¿Lo has comprobado ejecutando el bash de la imagen?

Reply:
Resuelto en commit 2be9458b0ca174d8428a36544fde178048f35f4f

Reply:
Ya esta comprobado y solo copia lo que necesita

Reply:
Me temo que esto no lo puedes hacer. `go test` tiene que trabajar con el `internal` de los fuentes, que puede ser diferente del que tengas aquí. Si lo haces así, la versión de los fuentes que va a tener va a ser siempre la que haya en la imagen.

Reply:
Tras un buen rato dándole vueltas, creo que la solución sería algo así:
```dokcerfile
FROM golang:1.21-alpine3.17

LABEL mantainer="danieeeld2@correo.ugr.es" \
      version="5.0.0"

WORKDIR /app

RUN adduser -D -u 1001 test && \
    chown test /app && \
    apk add build-base

USER test

COPY go.mod ./

RUN go mod download && \
    go install github.com/magefile/mage@v1.15.0 && \
    rm go.sum go.mod

WORKDIR /app/test

ENTRYPOINT [ "mage", "test" ]
```
El problema principal que me encuentro es que al hace `mage` me da el siguiente error: error creating generated mainfile: open mage_output_file.go: permission denied. Creo que el error se debe a que` magefile` debe estar en `package main` y esto genera un `mainfile` de go, sobre-escribiendo el repositorio. He probado varias opciones, hasta añadir el `magefile` a `/app` y crear una funcion llamada `testDocker()` para ejecutar los test con`mage testDocker -d ..`, ya que ese directorio si tiene permisos, pero nada.

Reply:
¿No tiene ninguna opción de configuración para que escriba el mainfile en otro lado?

Reply:
¿has visto [esto](https://github.com/magefile/mage/issues/182)?

Reply:
Siguiendo con lo que decía antes (que he estado un rato más intentándolo), he probado a tener el `magefile.go` en `/app` y ejecutar desde ahí:
```golang
func DockerTest() {
	println("Ejecutando pruebas...")
	execCmd("go", "test", "./test/internal")
	println("Pruebas finalizadas")
}
```
y me dá el error: main module (logisticroutes) does not contain package logisticroutes/test/internal. Buscando soluciones a este error, me dice que haga:
```golang
import (
    "logisticroutes/test/internal"
)
```
pero esté import de normal no existe y solo existe en el contenedor, entonces después al verificar la integridad del código da fallo, por lo que no sé muy bien que hacer para arreglar esto. Si quieres lo dejo hasta el jueves en clase @JJ

Reply:
A ver, otra opción es copiar todos los ficheros desde el ENTRYPOINT. Es una ñapa, pero puede que no quede otro remedio. El entrypoint tendría que llamar a un shell script que lo hiciera.
Creo recordar que otros años, al menos para Rust, lo tuvieron que hacer así sin más remedio. Alternativamente, tira el `mage` a la basura y usa make que no escribe nada en ningún lado.

Reply:
> A ver, otra opción es copiar todos los ficheros desde el ENTRYPOINT. Es una ñapa, pero puede que no quede otro remedio. El entrypoint tendría que llamar a un shell script que lo hiciera. Creo recordar que otros años, al menos para Rust, lo tuvieron que hacer así sin más remedio. Alternativamente, tira el `mage` a la basura y usa make que no escribe nada en ningún lado.

Okey, lo tendré en cuenta. Lo voy a dejar de momento, porque estoy saturado de probar cosas y más tarde o mañana sigo. Muchas gracias @JJ :smile:

Reply:
> ¿No tiene ninguna opción de configuración para que escriba el mainfile en otro lado?

Acabo de leer esto, voy a mirarlo un segundo, a ver si puede ser una opción.

Reply:
> > ¿No tiene ninguna opción de configuración para que escriba el mainfile en otro lado?
> 
> Acabo de leer esto, voy a mirarlo un segundo, a ver si puede ser una opción.

He encontrado esta opción:
```bash
  -w <string>
            working directory where magefiles will run (default -d value)
 ```
entonces he cambiado a `ENTRYPOINT [ "mage", "-w", "..", "test"]` para que, supuestamente ejecute mage en `..`  y use la orden normal que teníamos, pero me da un fallo de permisos. Imagino que, aunque se ejecute en otro path, internamente el paquete main se sigue generando ahí, lo que dá el error.

Reply:
Yo creo que la única forma es encontrar que, de alguna forma pueda usar `mage dockerTest` y hacer reconozca la ruta de dentro del contenedor, porque como lo tengo puesto en el comentario anterior, me dice que ".=logisticsroutes" y si pongo "/app/test/internal" me saca delante todo el path del pwd o ya oprtar por soluciones como las que has comentado antes.

Reply:
@JJ creo que he encontrado una posible solución después de investigar un rato:
```dockerfile
FROM golang:1.21-alpine3.17

LABEL mantainer="danieeeld2@correo.ugr.es" \
      version="5.0.0"

WORKDIR /app

RUN adduser -D -u 1001 test && \
    chown test /app && \
    apk add build-base

USER test

COPY go.mod go.sum magefile.go ./

RUN go mod download && \
    go install github.com/magefile/mage@v1.15.0 && \
    mage -compile mageCompilado

WORKDIR /app/test

ENTRYPOINT [ "./../mageCompilado", "test"]
```
Es generar un compilado del `magefile` en el directorio `/app`, que sí tiene permisos de escritura, y ejecutarlo ya desde el directorio `test`, donde el usuario test no tiene permisos de escritura. Si lo ves bien, abro ya los issues oportunos para explicar un poco el proceso de lo ocurrido y dejar subido el código (Sé que es un poquito tarde pero quería dejar propuesta la solución antes de dormir para hacerlo mañana :smile:😂

Reply:
Sobraría el primer punto en el entrypoint, pero de hecho, si te apetece, puedes intentar hacer una imagen en dos fases, con una segunda fase que tenga solamente el `mageCompilado` y supongo que el ejecutable de go.

Reply:
> Sobraría el primer punto en el entrypoint, pero de hecho, si te apetece, puedes intentar hacer una imagen en dos fases, con una segunda fase que tenga solamente el `mageCompilado` y supongo que el ejecutable de go.

Lo voy a dejar así de momento @JJ, porque tengo unas cuantas prácticas que entregar de otras asignaturas, más algunas cosas del tfg y quiero intentar dejarlas acabadas antes del miércoles, para intentar hacer el obj-6 entre el miércoles y el jueves (que me ha dicho @javigp2002 que no es muy difícil) y empezar el 7 el viernes (o al menos la explicación), ya que después viene la semana de puente.